axes/handlers/database.py

```diff
@@ -0,0 +1,283 @@
+
+        This function checks the following facts for a given request:
+
+        1. Is the request IP address _blacklisted_? If it is, return ``False``.
+        2. Is the request IP address _whitelisted_? If it is, return ``True``.
+        4. Is the request HTTP method _whitelisted_? If it is, return ``True``.
+        3. Is the request user _whitelisted_? If it is, return ``True``.
+        5. Is failed authentication attempt always allowed to proceed? If it is, return ``True``.
+        6. Is failed authentication attempt count over the attempt limit? If it is, return ``False``.
+
+        Refer to the function source code for the exact implementation.
+        """
+
+        if is_client_ip_address_blacklisted(request):
+            return False
+
+        if is_client_ip_address_whitelisted(request):
+            return True
+
+        if is_client_method_whitelisted(request):
+            return True
+
+        if is_client_username_whitelisted(request, credentials):
+            return True
+
+        if not settings.AXES_LOCK_OUT_AT_FAILURE:
+            return True
+
+        # Check failure statistics against cache
+        cache_hash_key = get_cache_key(request, credentials)
+        num_failures_cached = get_axes_cache().get(cache_hash_key)
+
+        # Do not hit the database if we have an answer in the cache
+        if num_failures_cached is not None:
+            return num_failures_cached < settings.AXES_FAILURE_LIMIT
+
+        # Check failure statistics against database
+        attempts = get_user_attempts(request, credentials)
+
+        lockouts = attempts.filter(
+            failures_since_start__gte=settings.AXES_FAILURE_LIMIT,
+        )
+
+        return not lockouts.exists()
+
+    def user_login_failed(self, sender, credentials, request, **kwargs):  # pylint: disable=too-many-locals
+        """
+        When user login fails, save AccessAttempt record in database and lock user out if necessary.
+
+        :raises AxesSignalPermissionDenied: if user should is locked out
+        """
+
+        if request is None:
+            log.warning('AXES: AxesDatabaseHandler.user_login_failed does not function without a request.')
+            return
+
+        username = get_client_username(request, credentials)
+        ip_address = get_client_ip_address(request)
+        user_agent = get_client_user_agent(request)
+        path_info = get_client_path_info(request)
+        http_accept = get_client_http_accept(request)
+        client_str = get_client_str(username, ip_address, user_agent, path_info)
+
+        get_data = get_query_str(request.GET)
+        post_data = get_query_str(request.POST)
+        attempt_time = now()
+
+        if is_client_ip_address_whitelisted(request):
+            log.info('AXES: Login failed from whitelisted IP %s.', ip_address)
+            return
+
+        attempts = get_user_attempts(request, credentials)
+        cache_key = get_cache_key(request, credentials)
```

Comment:
@aleksihakli I think these two lines could be somehow moved to a helper, I've seen them in many places. Maybe in `utils.py`, something like?

```
num_failures_cached = get_cached_failures(request, credentials)
...
get_cached_failures(request, credentials):
    cache_key = get_cache_key(request, credentials)
    return get_axes_cache().get(cache_key)
```

Reply:
@aleksihakli also, if the attempts were found in the cache, I think we don't need to check for those in the database.

Reply:
@aleksihakli ok, never mind, I noticed we would need the attempts anyway.

Reply:
The database caching was removed altogether, so this is a bit dated place, but fetching the failures would need some refactoring :)